File: src/main/scala/at/pria/koza/harmonic/BranchListener.java
/**
 * BranchListener.java
 * 
 * Created on 07.08.2013
 */

package at.pria.koza.harmonic;


import java.util.EventListener;


/**
 * <p>
 * {@code BranchListener} provides a listener interface that can be used to detect when a {@link BranchManager}'s
 * branch is created or moved.
 * </p>
 * 
 * @version V1.0 07.08.2013
 * @author SillyFreak
 */
public interface BranchListener extends EventListener {
    /**
     * <p>
     * Called when a new branch is created, either manually or by receiving a branch update.
     * </p>
     * 
     * @param mgr the BranchManager a branch is added to
     * @param branch the branch being created
     * @param head the initial tip of the new branch
     */
    public void branchCreated(BranchManager mgr, String branch, State head);
    
    /**
     * <p>
     * Called when an existing branch is moved, either manually or by receiving a branch update.
     * </p>
     * 
     * @param mgr the BranchManager the branch belongs to
     * @param branch the branch being moved
     * @param prevHead the previous tip of the branch
     * @param newHead the new tip of the branch
     */
    public void branchMoved(BranchManager mgr, String branch, State prevHead, State newHead);
    
    /**
     * <p>
     * Called when a branch is deleted.
     * </p>
     * 
     * @param mgr the BranchManager a branch is deleted from
     * @param branch the branch being deleted
     * @param prevHead the tip of the new branch previous to deleting it
     */
    public void branchDeleted(BranchManager mgr, String branch, State prevHead);
}

File: src/main/scala/at/pria/koza/harmonic/BranchListener.scala
/**
 * BranchListener.scala
 *
 * Created on 07.08.2013
 */

package at.pria.koza.harmonic

import java.util.EventListener

/**
 * <p>
 * `BranchListener`s listen to branches being created, moved, or added in an engine.
 * </p>
 *
 * @version V1.0 07.08.2013
 * @author SillyFreak
 */
trait BranchListener extends EventListener {
  /**
   * <p>
   * Called when a new branch is added to the engine.
   * </p>
   *
   * @param engine the Engine a branch is added to
   * @param branch the branch being created
   * @param tip the initial tip of the new branch
   */
  def branchCreated(engine: Engine, branch: String, tip: State): Unit = {}

  /**
   * <p>
   * Called when an existing branch is moved.
   * </p>
   *
   * @param engine the Engine containing the branch
   * @param branch the branch being moved
   * @param prevTip the previous tip of the branch
   * @param newTip the new tip of the branch
   */
  def branchMoved(engine: Engine, branch: String, prevTip: State, newTip: State): Unit = {}

  /**
   * <p>
   * Called when a branch is deleted.
   * </p>
   *
   * @param engine the Engine the branch is deleted from
   * @param branch the branch being deleted
   * @param tip the tip of the branch previous to being deleted
   */
  def branchDeleted(engine: Engine, branch: String, tip: State): Unit = {}
}
